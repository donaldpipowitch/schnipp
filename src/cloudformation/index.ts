import {
  CloudFormationCustomResourceEvent,
  CloudFormationCustomResourceResponse
} from 'aws-lambda';
import { IHandler } from '../';

export type CloudFormationCustomResourceHandler = IHandler<
  CloudFormationCustomResourceEvent,
  CloudFormationCustomResourceResponse
>;
