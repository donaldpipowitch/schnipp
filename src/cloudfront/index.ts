import {
  CloudFrontRequestEvent,
  CloudFrontRequestResult,
  CloudFrontResponseEvent,
  CloudFrontResponseResult
} from 'aws-lambda';
import { IHandler } from '..';

export type CloudFrontRequestHandler = IHandler<
  CloudFrontRequestEvent,
  CloudFrontRequestResult
>;

export type CloudFrontResponseHandler = IHandler<
  CloudFrontResponseEvent,
  CloudFrontResponseResult
>;
