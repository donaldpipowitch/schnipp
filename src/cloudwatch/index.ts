import { CloudWatchLogsEvent } from 'aws-lambda';
import { IHandler } from '../';

export type CloudwatchHandler = IHandler<CloudWatchLogsEvent, void>;
