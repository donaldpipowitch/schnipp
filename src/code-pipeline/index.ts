import { IHandler } from '..';
import { CodePipelineEvent } from 'aws-lambda';

export type CodePipelineHandler = IHandler<CodePipelineEvent>;
