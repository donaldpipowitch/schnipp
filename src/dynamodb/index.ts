import { DynamoDBStreamEvent } from 'aws-lambda';
import { IHandler } from '../';

export type DynamoDBStreamHandler = IHandler<DynamoDBStreamEvent, void>;
