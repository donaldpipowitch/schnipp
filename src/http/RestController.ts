import { APIGatewayHandler, createResponse } from '.';
import { CONTAINER_INSTANCE_PROP, Container, Newable } from 'simple-ts-di';
import {
  APIGatewayEvent,
  Context,
  APIGatewayProxyResult,
  APIGatewayProxyEvent
} from 'aws-lambda';
import {
  findMethod,
  MethodMetaData,
  DecoratedExceptionFilter,
  ExceptionFilter,
  Guard,
  MetaData
} from './factories/createHttpMethodDecorator';
import { HttpException } from './exceptions/HttpException';
import { ForbiddenException } from './exceptions/ForbiddenException';
import { getControllerMetaData } from './factories/createClassDectorator';

export class RestController implements APIGatewayHandler {
  private controllerData = getControllerMetaData(this.constructor);

  private get container() {
    return (this as any)[CONTAINER_INSTANCE_PROP] as Container;
  }

  async handle(
    event: APIGatewayProxyEvent,
    context: Context
  ): Promise<APIGatewayProxyResult> {
    const method = findMethod(
      this.controllerData.basePath,
      event.resource,
      event.httpMethod,
      this
    );

    if (method == null) {
      return createResponse(404);
    }

    const guards = [...method.guards, ...this.controllerData.guards];
    const metaData = { ...this.controllerData.metaData, ...method.metaData };

    try {
      await this.processGuards(guards, event, context, metaData);

      const headers = { ...this.controllerData.headers, ...method.headers };

      return createResponse(
        method.statusCode!,
        await this.generateResponse(method, event, context),
        headers
      );
    } catch (exception) {
      const filters = [...method.filters, ...this.controllerData.filters];

      const mappedExceptionResponse = await this.processFilters(
        filters,
        exception,
        event,
        context,
        metaData
      );

      if (mappedExceptionResponse != null) {
        return mappedExceptionResponse;
      }

      if (exception instanceof HttpException) {
        return createResponse(exception.statusCode, {
          statusCode: exception.statusCode,
          message: exception.message
        });
      }

      return createResponse(500, {
        statusCode: 500,
        message: 'Internal Server Error'
      });
    }
  }

  private async generateResponse(
    method: MethodMetaData,
    event: APIGatewayEvent,
    context: Context
  ) {
    return await (this as any)[method.propertyKey].apply(
      this,
      method.paramaters
        .sort((a, b) => a.index - b.index)
        .map(r => r.injector(event, method.metaData, context))
    );
  }

  private async processFilters(
    filters: Array<DecoratedExceptionFilter>,
    e: any,
    event: APIGatewayEvent,
    context: Context,
    metaData: MetaData
  ) {
    try {
      const exceptionFilter = filters.find(x =>
        x.exceptions.some(y => e.name === y.name || e instanceof y)
      );

      if (exceptionFilter == null) {
        return;
      }

      const filter = await this.container.get<ExceptionFilter>(
        exceptionFilter.filter
      );

      return await filter.catch(e, event, metaData, context);
    } catch (e) {
      return createResponse(500, { message: 'Internal Server Error' });
    }
  }

  private async processGuards(
    guards: Array<Newable<Guard>>,
    event: APIGatewayEvent,
    context: Context,
    metaData: MetaData
  ) {
    const canActivate = (await Promise.all(
      guards.map(guard =>
        this.container
          .get(guard)
          .then(guard => guard.canActivate(event, metaData, context))
      )
    )).every(Boolean);

    if (!canActivate) {
      throw new ForbiddenException();
    }
  }
}
