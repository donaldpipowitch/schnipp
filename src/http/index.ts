import {
  APIGatewayEvent,
  APIGatewayProxyResult,
  CustomAuthorizerEvent,
  CustomAuthorizerResult
} from 'aws-lambda';
import { IHandler } from '..';

export type APIGatewayHandler = IHandler<
  APIGatewayEvent,
  APIGatewayProxyResult
>;

export type AuthorizeHandler = IHandler<
  CustomAuthorizerEvent,
  CustomAuthorizerResult
>;

export { RestController } from './RestController';

export {
  InjectedParam,
  MetaData,
  MethodMetaData,
  ParamInjector,
  DecoratedExceptionFilter,
  Guard,
  ExceptionFilter
} from './factories/createHttpMethodDecorator';

export { Catch } from './decorators/catch';
export {
  GET,
  POST,
  DELETE,
  PUT,
  PATCH,
  OPTIONS,
  HEAD
} from './decorators/methods';
export { Path } from './decorators/path';
export { Param } from './decorators/param';
export { Query } from './decorators/query';
export { Header } from './decorators/header';
export { StatusCode } from './decorators/statusCode';
export { AuthContext } from './decorators/authContext';
export { Meta } from './decorators/meta';
export { Body } from './decorators/body';
export { Use } from './decorators/use';

export { HttpException } from './exceptions/HttpException';
export { NotFoundException } from './exceptions/NotFoundException';
export { ForbiddenException } from './exceptions/ForbiddenException';

export { AuthPolicy, HttpVerb, Effect } from './utils/AuthPolicy';
export { createResponse } from './utils/createResponse';
