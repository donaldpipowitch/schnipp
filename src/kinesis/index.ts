import { KinesisStreamEvent } from 'aws-lambda';
import { IHandler } from '..';

export type KinesisStreamHandler = IHandler<KinesisStreamEvent>;
