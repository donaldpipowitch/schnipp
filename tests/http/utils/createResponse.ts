import { createResponse } from '../../../src/http';

describe('createResponse', () => {
  it('should create a proper response', () => {
    expect(createResponse(200)).toMatchSnapshot();
    expect(createResponse(400, { foo: 'bar' })).toMatchSnapshot();
    expect(
      createResponse(400, { foo: 'bar' }, { Accept: 'text/html' })
    ).toMatchSnapshot();
    expect(
      createResponse(300, 'base64', { Accept: 'text/html' })
    ).toMatchSnapshot();
  });
});
