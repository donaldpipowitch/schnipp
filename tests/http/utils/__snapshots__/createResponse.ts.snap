// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createResponse should create a proper response 1`] = `
Object {
  "body": "{}",
  "headers": Object {},
  "isBase64Encoded": false,
  "statusCode": 200,
}
`;

exports[`createResponse should create a proper response 2`] = `
Object {
  "body": "{\\"foo\\":\\"bar\\"}",
  "headers": Object {},
  "isBase64Encoded": false,
  "statusCode": 400,
}
`;

exports[`createResponse should create a proper response 3`] = `
Object {
  "body": "{\\"foo\\":\\"bar\\"}",
  "headers": Object {
    "Accept": "text/html",
  },
  "isBase64Encoded": false,
  "statusCode": 400,
}
`;

exports[`createResponse should create a proper response 4`] = `
Object {
  "body": "base64",
  "headers": Object {
    "Accept": "text/html",
  },
  "isBase64Encoded": true,
  "statusCode": 300,
}
`;
